go/chat/sender.go

```diff
@@ -0,0 +1,329 @@
+	uid := s.G().Env.GetUID()
+	if uid.IsNil() {
+		return chat1.MessagePlaintext{}, libkb.LoginRequiredError{}
+	}
+	did := s.G().Env.GetDeviceID()
+	if did.IsNil() {
+		return chat1.MessagePlaintext{}, libkb.DeviceRequiredError{}
+	}
+
+	huid := uid.ToBytes()
+	if huid == nil {
+		return chat1.MessagePlaintext{}, errors.New("invalid UID")
+	}
+
+	hdid := make([]byte, libkb.DeviceIDLen)
+	if err := did.ToBytes(hdid); err != nil {
+		return chat1.MessagePlaintext{}, err
+	}
+
+	header := msg.ClientHeader
+	header.Sender = gregor1.UID(huid)
+	header.SenderDevice = gregor1.DeviceID(hdid)
+	updated := chat1.MessagePlaintext{
+		ClientHeader: header,
+		MessageBody:  msg.MessageBody,
+	}
+	return updated, nil
+}
+
+func (s *BlockingSender) addPrevPointersToMessage(msg chat1.MessagePlaintext, convID chat1.ConversationID) (chat1.MessagePlaintext, error) {
+	// Make sure the caller hasn't already assembled this list. For now, this
+	// should never happen, and we'll return an error just in case we make a
+	// mistake in the future. But if there's some use case in the future where
+	// a caller wants to specify custom prevs, we can relax this.
+	if len(msg.ClientHeader.Prev) != 0 {
+		return chat1.MessagePlaintext{}, fmt.Errorf("chatLocalHandler expects an empty prev list")
+	}
+
+	res, _, err := s.G().ConvSource.Pull(context.Background(), convID, msg.ClientHeader.Sender, nil, nil)
+	if err != nil {
+		return chat1.MessagePlaintext{}, err
+	}
+
+	prevs, err := CheckPrevPointersAndGetUnpreved(&res)
+	if err != nil {
+		return chat1.MessagePlaintext{}, err
+	}
+
+	// Make an attempt to avoid changing anything in the input message. There
+	// are a lot of shared pointers though, so this is
+	header := msg.ClientHeader
+	header.Prev = prevs
+	updated := chat1.MessagePlaintext{
+		ClientHeader: header,
+		MessageBody:  msg.MessageBody,
+	}
+	return updated, nil
+}
+
+func (s *BlockingSender) Prepare(ctx context.Context, plaintext chat1.MessagePlaintext, convID *chat1.ConversationID) (*chat1.MessageBoxed, error) {
+	msg, err := s.addSenderToMessage(plaintext)
+	if err != nil {
+		return nil, err
+	}
+
+	// convID will be nil in makeFirstMessage, for example
+	if convID != nil {
+		msg, err = s.addPrevPointersToMessage(msg, *convID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// encrypt the message
+	skp, err := s.getSigningKeyPair()
+	if err != nil {
+		return nil, err
+	}
+
+	// For now, BoxMessage canonicalizes the TLF name. We should try to refactor
+	// it a bit to do it here.
+	boxed, err := s.boxer.BoxMessage(ctx, msg, skp)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: populate plaintext.ClientHeader.Conv
+
+	return boxed, nil
+}
+
+func (s *BlockingSender) getSigningKeyPair() (kp libkb.NaclSigningKeyPair, err error) {
+	// get device signing key for this user
+	signingKey, err := engine.GetMySecretKey(s.G(), s.getSecretUI, libkb.DeviceSigningKeyType, "sign chat message")
```

Comment:
What will happen if we actually need a passphrase from the background?  There's nothing possible to do here, is there?